project/src/test/java/com/google/sps/EventRankerServletTest.java

```diff
@@ -0,0 +1,130 @@
+import com.google.cloud.Date;
+import com.google.sps.data.Event;
+import com.google.sps.data.User;
+import com.google.sps.data.VolunteeringOpportunity;
+import com.google.sps.servlets.EventRankerServlet;
+import com.google.sps.utilities.CommonUtils;
+import com.google.sps.utilities.EventRanker;
+import com.google.sps.utilities.SpannerClient;
+import com.google.sps.utilities.SpannerTasks;
+import com.google.sps.utilities.SpannerTestTasks;
+import com.google.sps.utilities.TestUtils;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.servlet.ServletContextEvent;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mockito;
+import org.springframework.mock.web.MockServletContext;
+
+/** Tests the ranking algorithm for event discovery */
+@RunWith(JUnit4.class)
+public final class EventRankerServletTest {
+  private static final HttpServletRequest request = Mockito.mock(HttpServletRequest.class);
+  private static final HttpServletResponse response = Mockito.mock(HttpServletResponse.class);
+  private static final StringWriter stringWriter = new StringWriter();
+  private static final PrintWriter writer = new PrintWriter(stringWriter);
+  private static final LocalServiceTestHelper authenticationHelper =
+      new LocalServiceTestHelper(new LocalUserServiceTestConfig());
+  private static final String FOOD = "food";
+  private static final String SEWING = "sewing";
+  private static Event EVENT_FOOD;
+  private static Event EVENT_SEWING;
+  private static User USER;
+  private static String EMAIL = "test@example.com";
+  private static String INVALID_EMAIL = "invalid@example.com";
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    setUpDatabase();
+    authenticationHelper.setUp();
+    setUpEventsAndUsers();
+    Mockito.when(response.getWriter()).thenReturn(writer);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
+    SpannerTestTasks.cleanup();
+    authenticationHelper.tearDown();
+  }
+
+  @After
+  public void flushWriter() {
+    Mockito.reset(request);
+    stringWriter.getBuffer().setLength(0);
+  }
+
+  @Test
+  public void testServletGetRanking() throws IOException {
+    setAuthenticationHelper(EMAIL);
+    SpannerTasks.insertOrUpdateUser(USER);
+    Set<Event> events = new HashSet<>(Arrays.asList(EVENT_FOOD, EVENT_SEWING));
+    for (Event event : events) {
+      SpannerTasks.insertorUpdateEvent(event);
+    }
+    List<Event> expectedEvents = EventRanker.rankEvents(USER, events);
```

Comment:
I understand why you use rankEvents here. But in general we don't want to make test to depend another code, since the test might false passed and failed due to another part of your production. Let's hardcode the expectedEvents to make the test not depend on EventRanker.

Reply:
Done - that makes sense, I thought that since we tested EventRanker elsewhere we could rely on it working in this test. But it would be more reliable to hardcode.